File: 20.有效的括号.js
/*
 * @lc app=leetcode.cn id=20 lang=javascript
 *
 * [20] 有效的括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let rightSymbols = []
  for (let i = 0; i < s.length; i++) {
    switch (s[i]) {
      case '(':
        rightSymbols.push(')')
        break
      case '[':
        rightSymbols.push(']')
        break
      case '{':
        rightSymbols.push('}')
        break
      default:
        if (rightSymbols.pop() !== s[i]) {
          return false
        }
    }
  }
  return !rightSymbols.length
}
// @lc code=end

File: 20.有效的括号.go
/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */

// @lc code=start
func isValid(s string) bool {
	hash := map[byte]byte{')': '(', ']': '[', '}': '{'}
	stack := make([]byte, 0)
	if s == "" {
		return true
	}

	for i := 0; i < len(s); i++ {
		if s[i] == '(' || s[i] == '[' || s[i] == '{' {
			stack = append(stack, s[i])
		} else if len(stack) > 0 && stack[len(stack)-1] == hash[s[i]] {
			stack = stack[:len(stack)-1]
		} else {
			return false
		}
	}
	return len(stack) == 0
}

// @lc code=end

